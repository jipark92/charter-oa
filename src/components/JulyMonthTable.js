import React from 'react'

export default function JulyMonthTable({ sortedJulyTable }) {
    return (
        <div>
            <h1>JUNE</h1>
            {/* customer 123 */}
            <table border="1">
                <thead>
                    <tr>
                        <th>Customer ID</th>
                        <th>Transaction ID</th>
                        <th>Date</th>
                        <th>Price</th>
                        <th>Rewards</th>
                    </tr>
                </thead>
                <tbody>
                    {sortedJulyTable().map((data, i) => {
                        if (data.customerID === 123) {
                            return (
                                <tr key={i}>
                                    <td>{data.customerID}</td>
                                    <td>{data.transactionID}</td>
                                    <td>{data.purchasedDate}</td>
                                    <td>${data.price}</td>
                                    <td>{data.rewardPoints}</td>
                                </tr>
                            )
                        }
                    })}
                    <tr>
                        <td colSpan="4"><b>TOTAL REWARDS</b></td>
                        <td><b>0</b></td>
                    </tr>
                </tbody>
            </table>

            {/* customer 456 */}
            <table border="1">
                <thead>
                    <tr>
                        <th>Customer ID</th>
                        <th>Transaction ID</th>
                        <th>Date</th>
                        <th>Price</th>
                        <th>Rewards</th>
                    </tr>
                </thead>
                <tbody>
                    {sortedJulyTable().map((data, i) => {
                        if (data.customerID === 456) {
                            return (
                                <tr key={i}>
                                    <td>{data.customerID}</td>
                                    <td>{data.transactionID}</td>
                                    <td>{data.purchasedDate}</td>
                                    <td>${data.price}</td>
                                    <td>{data.rewardPoints}</td>
                                </tr>
                            )
                        }
                    })}
                    <tr>
                        <td colSpan="4"><b>TOTAL REWARDS</b></td>
                        <td><b>0</b></td>
                    </tr>
                </tbody>
            </table>

            {/* customer 789 */}
            <table border="1">
                <thead>
                    <tr>
                        <th>Customer ID</th>
                        <th>Transaction ID</th>
                        <th>Date</th>
                        <th>Price</th>
                        <th>Rewards</th>
                    </tr>
                </thead>
                <tbody>
                    {sortedJulyTable().map((data, i) => {
                        if (data.customerID === 789) {
                            return (
                                <tr key={i}>
                                    <td>{data.customerID}</td>
                                    <td>{data.transactionID}</td>
                                    <td>{data.purchasedDate}</td>
                                    <td>${data.price}</td>
                                    <td>{data.rewardPoints}</td>
                                </tr>
                            )
                        }
                    })}
                    <tr>
                        <td colSpan="4"><b>TOTAL REWARDS</b></td>
                        <td><b>0</b></td>
                    </tr>
                </tbody>
            </table>
        </div>
    )
}
