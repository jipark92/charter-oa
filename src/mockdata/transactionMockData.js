export const calculateRewardPoints = (price) => {
    const moreThan50 = 1
    const moreThan100 = 2
    if (price > 100) {
        return (price - 100) * moreThan100 + 50 * moreThan50
    } else if (price > 50) {
        return (price - 50) * moreThan50
    } else {
        return 0
    }
}

const transactionMockData = [
    {
        transactionID: 1,
        customerID: 123,
        price: 100,
        rewardPoints: calculateRewardPoints(100),
        purchasedDate: new Date("July 01, 22 ").toDateString(),
    },
    {
        transactionID: 2,
        customerID: 456,
        price: 60,
        rewardPoints: calculateRewardPoints(60),
        purchasedDate: new Date("June 19, 22 ").toDateString(),
    },
    {
        transactionID: 3,
        customerID: 789,
        price: 175,
        rewardPoints: calculateRewardPoints(175),
        purchasedDate: new Date("May 23, 22 ").toDateString(),
    },
    {
        transactionID: 4,
        customerID: 123,
        price: 55,
        rewardPoints: calculateRewardPoints(55),
        purchasedDate: new Date("May 21, 22 ").toDateString(),
    },
    {
        transactionID: 5,
        customerID: 456,
        price: 125,
        rewardPoints: calculateRewardPoints(125),
        purchasedDate: new Date("May 11, 22 ").toDateString(),
    },
    //
    {
        transactionID: 6,
        customerID: 789,
        price: 100,
        rewardPoints: calculateRewardPoints(100),
        purchasedDate: new Date("June 25, 22 ").toDateString(),
    },
    {
        transactionID: 7,
        customerID: 123,
        price: 45,
        rewardPoints: calculateRewardPoints(45),
        purchasedDate: new Date("June 14, 22 ").toDateString(),
    },
    {
        transactionID: 8,
        customerID: 456,
        price: 203,
        rewardPoints: calculateRewardPoints(203),
        purchasedDate: new Date("July 15, 22 ").toDateString(),
    },
    {
        transactionID: 9,
        customerID: 789,
        price: 126,
        rewardPoints: calculateRewardPoints(126),
        purchasedDate: new Date("July 15, 22 ").toDateString(),
    },
    {
        transactionID: 10,
        customerID: 123,
        price: 500,
        rewardPoints: calculateRewardPoints(500),
        purchasedDate: new Date("May 14, 22 ").toDateString(),
    },
    {
        transactionID: 11,
        customerID: 456,
        price: 75,
        rewardPoints: calculateRewardPoints(75),
        purchasedDate: new Date("June 15, 22 ").toDateString(),
    },
    {
        transactionID: 12,
        customerID: 789,
        price: 89,
        rewardPoints: calculateRewardPoints(89),
        purchasedDate: new Date("July 15, 22 ").toDateString(),
    },
    {
        transactionID: 13,
        customerID: 123,
        price: 152,
        rewardPoints: calculateRewardPoints(152),
        purchasedDate: new Date("May 15, 22 ").toDateString(),
    },
];

export const fetchTransactionMockData = () => {
    return new Promise((res, rej) => {
        setTimeout(() => {
            res(transactionMockData)
        }, 1000);
    })
}