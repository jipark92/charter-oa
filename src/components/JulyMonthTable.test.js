import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { transactionMockData } from '../mockdata/transactionMockData'
import JulyMonthTable from './JulyMonthTable';
//categorized table array
const julyTable = []

//find july data
const findJulyDateAndCustomerID = (() => {
    transactionMockData.map((data) => {
        if (data.customerID && new Date(data.purchasedDate).getMonth() + 1 === 7) {
            julyTable.push(data)
        }
    })
})()
const sortedJulyTable = () => julyTable.sort((a, b) => new Date(a.purchasedDate) - new Date(b.purchasedDate))

//calcualte july total reward points
const calculateJulyTotalRewards = (customer) => {
    const julyRewardPointsArr = []
    julyTable.map((data) => {
        if (data.customerID === customer) {
            julyRewardPointsArr.push(data.rewardPoints)
        }
    })
    return julyRewardPointsArr.reduce((total, pointsArr) => {
        return total + pointsArr
    }, 0)
}
//test
describe(`JulyMonthTable should render customer's July information in tables`, () => {

    it('should match snapshot', () => {
        const { asFragment } = render(<JulyMonthTable
            sortedJulyTable={sortedJulyTable}
            calculateJulyTotalRewards={calculateJulyTotalRewards}
        />)
        expect(asFragment()).toMatchSnapshot()
    })

    it('header should be July', () => {
        render(<JulyMonthTable
            sortedJulyTable={sortedJulyTable}
            calculateJulyTotalRewards={calculateJulyTotalRewards}
        />)
        const headerJuly = screen.getByTestId('header-july').textContent
        expect(headerJuly).toBe('JULY')
    })

    it('there should be 3 TableHeaders in July because there are 3 customers', () => {
        render(<JulyMonthTable
            sortedJulyTable={sortedJulyTable}
            calculateJulyTotalRewards={calculateJulyTotalRewards}
        />)

        expect(screen.getAllByTestId('th-customerid').length).toEqual(3)
        expect(screen.getAllByTestId('th-transactionid').length).toEqual(3)
        expect(screen.getAllByTestId('th-date').length).toEqual(3)
        expect(screen.getAllByTestId('th-price').length).toEqual(3)
        expect(screen.getAllByTestId('th-rewards').length).toEqual(3)
    })

    it('total for customer 123 should be 1009', () => {
        render(<JulyMonthTable
            sortedJulyTable={sortedJulyTable}
            calculateJulyTotalRewards={calculateJulyTotalRewards}
        />)
        const customer123Total = screen.getByTestId('customer123Total').textContent
        expect(customer123Total).toBe('1630')
    })

    it('total for customer 456 should be 100', () => {
        render(<JulyMonthTable
            sortedJulyTable={sortedJulyTable}
            calculateJulyTotalRewards={calculateJulyTotalRewards}
        />)
        const customer456Total = screen.getByTestId('customer456Total').textContent
        expect(customer456Total).toBe('256')
    })

    it('total for customer 789 should be 2512', () => {
        render(<JulyMonthTable
            sortedJulyTable={sortedJulyTable}
            calculateJulyTotalRewards={calculateJulyTotalRewards}
        />)
        const customer789Total = screen.getByTestId('customer789Total').textContent
        expect(customer789Total).toBe('141')
    })


})