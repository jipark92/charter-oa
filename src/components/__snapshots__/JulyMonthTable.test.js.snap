// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JulyMonthTable should render customer's July information in tables should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1
      data-testid="header-july"
    >
      JULY
    </h1>
    <table
      border="1"
    >
      <thead>
        <tr>
          <th
            data-testid="th-customerid"
          >
            Customer ID
          </th>
          <th
            data-testid="th-transactionid"
          >
            Transaction ID
          </th>
          <th
            data-testid="th-date"
          >
            Date
          </th>
          <th
            data-testid="th-price"
          >
            Price
          </th>
          <th
            data-testid="th-rewards"
          >
            Rewards
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            123
          </td>
          <td>
            1
          </td>
          <td>
            Fri Jul 01 2022
          </td>
          <td>
            $890
          </td>
          <td>
            1630
          </td>
        </tr>
        <tr>
          <td
            colspan="4"
          >
            <b>
              TOTAL REWARDS
            </b>
          </td>
          <td
            data-testid="customer123Total"
          >
            <b>
              <i>
                1630
              </i>
            </b>
          </td>
        </tr>
      </tbody>
    </table>
    <table
      border="1"
    >
      <thead>
        <tr>
          <th
            data-testid="th-customerid"
          >
            Customer ID
          </th>
          <th
            data-testid="th-transactionid"
          >
            Transaction ID
          </th>
          <th
            data-testid="th-date"
          >
            Date
          </th>
          <th
            data-testid="th-price"
          >
            Price
          </th>
          <th
            data-testid="th-rewards"
          >
            Rewards
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            456
          </td>
          <td>
            8
          </td>
          <td>
            Fri Jul 15 2022
          </td>
          <td>
            $203
          </td>
          <td>
            256
          </td>
        </tr>
        <tr>
          <td
            colspan="4"
          >
            <b>
              TOTAL REWARDS
            </b>
          </td>
          <td
            data-testid="customer456Total"
          >
            <b>
              <i>
                256
              </i>
            </b>
          </td>
        </tr>
      </tbody>
    </table>
    <table
      border="1"
    >
      <thead>
        <tr>
          <th
            data-testid="th-customerid"
          >
            Customer ID
          </th>
          <th
            data-testid="th-transactionid"
          >
            Transaction ID
          </th>
          <th
            data-testid="th-date"
          >
            Date
          </th>
          <th
            data-testid="th-price"
          >
            Price
          </th>
          <th
            data-testid="th-rewards"
          >
            Rewards
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            789
          </td>
          <td>
            12
          </td>
          <td>
            Thu Jul 14 2022
          </td>
          <td>
            $89
          </td>
          <td>
            39
          </td>
        </tr>
        <tr>
          <td>
            789
          </td>
          <td>
            9
          </td>
          <td>
            Mon Jul 25 2022
          </td>
          <td>
            $126
          </td>
          <td>
            102
          </td>
        </tr>
        <tr>
          <td
            colspan="4"
          >
            <b>
              TOTAL REWARDS
            </b>
          </td>
          <td
            data-testid="customer789Total"
          >
            <b>
              <i>
                141
              </i>
            </b>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
